provider/azure/environ.go

```diff
@@ -1074,7 +1074,56 @@ func (env *azureEnviron) instances(
 
 // AdoptResources is part of the Environ interface.
 func (env *azureEnviron) AdoptResources(controllerUUID string, fromVersion version.Number) error {
-	return errors.NotImplementedf("AdoptResources")
+	groupClient := resources.GroupsClient{env.resources}
+	resourceClient := resources.Client{env.resources}
+	var failed []string
+
+	resources, err := groupClient.ListResources(env.resourceGroup, "", nil)
+	if err != nil {
+		return errors.Annotate(err, "listing resources")
+	}
+	for resources.Value != nil {
+		for _, resource := range *resources.Value {
+
+			providerNamespace, parentPath, err := splitResourceType(*resource.Type)
+			if err != nil {
+				failed = append(failed, *resource.Name)
+				logger.Errorf("error getting namespace and path for type %q: %v", *resource.Type, err)
+			}
+
+			(*resource.Tags)[tags.JujuController] = &controllerUUID
```

Comment:
Oh, I see! Yes, it can get restarted after crashing halfway through - the migration master will retry the phase if it errors out, so this would be called again on the half-updated group. 

I think it would always handle that right (although your comment here will make it more efficient). We're finding the resources by group, based on the model id which hasn't changed. Some of the resources will have the old controller tag and some will have the new one, and they'll all be updated to the new one (barring more network errors).